src/engine/Server.cpp

```diff
@@ -529,39 +361,57 @@ Awaitable<void> Server::process(
     }
   }
 
-  auto visitQuery = [&checkParameter, &accessTokenOk, &request, &send,
-                     &parameters, &requestTimer,
-                     this](Query query) -> Awaitable<void> {
+  auto visitOperation = [&checkParameter, &accessTokenOk, &request, &send,
+                         &parameters, &requestTimer,
+                         this]<QL_CONCEPT_OR_TYPENAME(QueryOrUpdate) Operation>(
+                            const Operation& op, auto opFieldString,
+                            std::function<bool(const ParsedQuery&)> pred,
+                            std::string msg) -> Awaitable<void> {
     if (auto timeLimit = co_await verifyUserSubmittedQueryTimeout(
             checkParameter("timeout", std::nullopt), accessTokenOk, request,
             send)) {
-      co_return co_await processQueryOrUpdate(parameters, query, requestTimer,
-                                              std::move(request), send,
-                                              timeLimit.value());
+      ad_utility::websocket::MessageSender messageSender = createMessageSender(
+          queryHub_, request, std::invoke(opFieldString, op));
+      auto [parsedOperation, qec, cancellationHandle,
+            cancelTimeoutOnDestruction] =
+          parseOperation(messageSender, parameters, op, timeLimit.value());
+      if (pred(parsedOperation)) {
+        throw std::runtime_error(
+            absl::StrCat(msg, parsedOperation._originalString));
+      }
+      if constexpr (std::is_same_v<Operation, Query>) {
+        co_return co_await processQuery(parameters, std::move(parsedOperation),
+                                        requestTimer, cancellationHandle, qec,
+                                        std::move(request), send,
+                                        timeLimit.value());
+      } else {
+        static_assert(std::is_same_v<Operation, Update>);
+        co_return co_await processUpdate(
+            std::move(parsedOperation), requestTimer, cancellationHandle, qec,
+            std::move(request), send, timeLimit.value());
+      }
     } else {
       // If the optional is empty, this indicates an error response has been
       // sent to the client already. We can stop here.
       co_return;
```

Comment:
Can you invert the control flow? First check the empty case + co_return, then the "normal" branch doesn't have to be nested.